### ISSUE TITLE
Update mods to not rely on roof adding magic

### ISSUE THREAD

**Issue by PatrikLundell** [CleverRaven/Cataclysm-DDA#74002]
### Is your feature request related to a problem? Please describe.

I'm trying to update the game's mapgen to explicitly add roofs where they should be in order to be able to remove the usage of roof adding magic.
However, I don't intend to deal with mods (and I won't mind if others deal with non mod instances as well), so in order for mods not to be somewhat broken when that happens (if it does, and many months from now), conversion ought to be performed.

### Solution you would like.

Mod maintainers going through their mods to ensure any mapgen they engage in produces all the roofs they should have.

### Describe alternatives you have considered.

Give no warning and let mods break. That would be rude.

### Additional context

An easy way to check for issues is to hack map.cpp operation map::add_roofs to return immediately (that disables the magic roof adding) and then debug spawn the construction (or map special) somewhere and teleport to look at them.
Debug spawning scaffolding works to provide a vantage point above the buildings, although teleporting onto where the roof should be would work as well.

The most common issue is to not providing a roof on top of the top story of a building.

Context:
What is "roof adding magic"?
That's my derisive name for the add_roofs functionality. It scans maps on loading and adds "roof" on top of any "terrain" that has a "roof" specified where there isn't a "roof" above it already. I want that usage gone, except for usage with trees, because it's not intuitive and causes issues (such as the "repair" of "roofs" above rooms you've just dug into). If you want a roof on top of your wall you should build it (and you should be allowed to build whatever roof you want (such as a wall for an additional level) rather than automatically get what the wall specified as its roof.
The exception for trees is because I don't see any good way to place them using existing methods, since their position should be somewhat random and you'd need to coordinate the tree top placement with the trunk.

**Comment by John-Candlebury** [comment]
There are a handfull of use cases where the "roof magic is desireable" the big one is generating ruined surface structures, as happens with aftershock formless ruins.

Those are completely chunk based and having to ensure each chunk is placed with a corresponding maching roof chunk above would be incredibly complex and make expanding them harder.

**Comment by PatrikLundell** [comment]
That's very useful information.

I'll try to look at this case to understand the case in detail, but I may come back later to ask for more info if I can't understand it.

**Comment by PatrikLundell** [comment]
I've had a first look at the aftershock formless ruins, but will have to look at it more.

- I didn't know you CAN specify chunks using the "rows" format. I've lamented it wasn't possible, which obviously is incorrect. That's definitely useful.
- I think it should be possible to use the nesting logic to make use of roof chunks, but I'll have to experiment by hacking an expansion. If it works, it should be possible to use for mapgen if mapgen supported 3D.

I'll be back after more investigations.

**Comment by PatrikLundell** [comment]
Master now supports 3D mapgen, so the formless ruins should now be possible to update with relative ease. The FEMA camp has a similar logic which has been updated to provide roofs, so it could be used as a reference, if needed.

**Comment by Maleclypse** [comment]
Ok so I've inherited some mapgen from vanilla `cave_rat` I think I need to work around the old roof magic.  I'll keep looking for others  that might break

**Comment by Maleclypse** [comment]
The genius locii maps use roof magic, I might let them just have open roofs though.  @Standing-Storm  will that make paraclesian starts more dangerous due to exposure?

**Comment by PatrikLundell** [comment]
I can help with updating mod maps if provided with info on what to change (@John-Candlebury has provided such info for Aftershock), but I can't perform the testing if them, as that requires knowledge of how to use the mods.

**Comment by Maleclypse** [comment]
> I can help with updating mod maps if provided with info on what to change (@John-Candlebury has provided such info for Aftershock), but I can't perform the testing if them, as that requires knowledge of how to use the mods.

Yeah the cave_rat needs whatever was done to other caves, I’d have to look at them to know. Are caves on z-1 or z0? If z0 I guess they just need a z1roof. Sorry I really need to look at caves to see how they work. I think they are on z0?

**Comment by PatrikLundell** [comment]
I just looked at the rat cave, and as far as I can see it has an ordinary cave at Z=0 and replaces Z=-1 with its own special version, and adds another special version to Z=-2.

I just debug spawned a rat cave, teleported to it, and looked at it, and as far as I can see it's fine. There isn't a hole in the roof where the slope down is (as I haven't hacked the code I'm playing with to suppress roof magic), which indicates that the normal cave code takes care of it.
